bundle/terranova/graph.go

```diff
@@ -0,0 +1,185 @@
+			}
+
+			for _, fieldRef := range fieldRefs {
+				for _, referencedNode := range fieldRef.referencedNodes {
+					label := fmt.Sprintf("%s.%s -> %s.%s", referencedNode.Group, referencedNode.Name, n.Group, n.Name)
+					log.Debugf(ctx, "Adding resource edge: %s (via %#v)", label, fieldRef.ref.Str)
+					g.AddDirectedEdge(
+						referencedNode,
+						n,
+						label,
+					)
+					isReferenced[referencedNode] = true
+				}
+			}
+			return dyn.InvalidValue, nil
+		},
+	)
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading config: %w", err)
+	}
+
+	return g, isReferenced, nil
+}
+
+func extractReferences(root dyn.Value, node nodeKey) ([]fieldRef, error) {
+	var result []fieldRef
+
+	val, err := dyn.GetByPath(root, dyn.NewPath(dyn.Key("resources"), dyn.Key(node.Group), dyn.Key(node.Name)))
+	if err != nil {
+		return nil, err
+	}
+
+	err = dyn.WalkReadOnly(val, func(p dyn.Path, v dyn.Value) error {
+		ref, ok := dynvar.NewRef(v)
+		if !ok {
+			return nil
+		}
+		referencedNodes, err := nodeFromRef(root, ref)
+		if err != nil {
+			return err
+		}
+		result = append(result, fieldRef{p, ref, referencedNodes})
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("parsing refs: %w", err)
+	}
+	return result, nil
+}
+
+func validateRef(root dyn.Value, ref string) (string, string, error) {
+	items := strings.Split(ref, ".")
+	if len(items) < 3 { // resources.jobs.foo.id
+		return "", "", errors.New("reference too short")
+	}
+	if items[0] != "resources" {
+		return "", "", errors.New("reference does not start with 'resources'")
+	}
+	_, err := dyn.GetByPath(root, dyn.NewPath(dyn.Key(items[0]), dyn.Key(items[1]), dyn.Key(items[2])))
+	if err != nil {
+		return "", "", err
+	}
+	if len(items) > 4 || items[3] != "id" {
+		return "", "", errors.New("${resources...} can only refer to field in the config or 'id'")
+	}
+	return items[1], items[2], nil
+}
+
+func nodeFromRef(root dyn.Value, ref dynvar.Ref) ([]nodeKey, error) {
+	var referencedNodes []nodeKey
+	for _, r := range ref.References() {
+		// validateRef will check resource exists in the config; this will reject references to deleted resources, no need to handle that case separately.
+		refGroup, refKey, err := validateRef(root, r)
+		if err != nil {
+			return nil, fmt.Errorf("cannot process reference %s: %w", r, err)
+		}
+		referencedNode := nodeKey{refGroup, refKey}
+		referencedNodes = append(referencedNodes, referencedNode)
+	}
+	return referencedNodes, nil
+}
+
+func resolveIDReference(ctx context.Context, b *bundle.Bundle, group, resourceName string) error {
+	mypath := dyn.NewPath(
+		dyn.Key("resources"),
+		dyn.Key(group),
+		dyn.Key(resourceName),
+		dyn.Key("id"),
+	)
+
+	entry, hasEntry := b.ResourceDatabase.GetResourceEntry(group, resourceName)
+	idValue := entry.ID
+	if !hasEntry || idValue == "" {
+		return errors.New("internal error: no db entry")
+	}
+
+	bundle.ApplyFuncContext(ctx, b, func(ctx context.Context, b *bundle.Bundle) {
```

Comment:
Why does it need to be inside `ApplyFuncContext` call and not just Config.Mutate?

Reply:
ApplyFuncContext ensures that dynamic configuration is copied back to typed one.